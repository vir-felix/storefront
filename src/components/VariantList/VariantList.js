import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import { withStyles } from "material-ui/styles";
import { observable, action, computed } from "mobx";
import { observer } from "mobx-react";

import { Router } from "routes";
import VariantItem from "components/VariantItem";
import ProductDetailOptionsList from "components/ProductDetailOptionsList";
import Divider from "components/Divider";

const styles = (theme) => ({
  variantsContainer: {
  },
  variantItem: {
    marginTop: theme.spacing.unit * 1.25,
    marginBottom: theme.spacing.unit * 1.25
  }
});

@withStyles(styles)
@observer
class VariantList extends Component {
  static propTypes = {
    classes: PropTypes.object,
    product: PropTypes.object.isRequired,
    variants: PropTypes.arrayOf(PropTypes.object).isRequired
  }

  constructor(props) {
    super(props);
    // Select first variant by default
    this.selectedVariant = props.variants[0]._id;
  }

  @observable _selectedVariant = null;

  @action
  handleClick = (variant) => {
    this._selectedVariant = variant._id;
    Router.pushRoute("product", {
      productSlug: this.props.product.slug,
      variantId: variant._id
    });
  }

  @computed
  get selectedVariant() {
    return this._selectedVariant;
  }

  set selectedVariant(value) {
    this._selectedVariant = value;
  }

  renderVariant = (variant) => {
    const { classes: { variantItem } } = this.props;
    const active = (this.selectedVariant === variant._id);

    return (
      <div className={variantItem} key={variant._id}>
        <VariantItem
          active={active}
          handleClick={this.handleClick}
          variant={variant}
        />
      </div>
    );
  }

  renderOptionsList = () => {
    const selectedVariant = this.props.variants.find((variant) => variant._id === this.selectedVariant);

    // If currently selected variant has options, then render them.
    const options = (selectedVariant.options.length) ? selectedVariant.options : null;

    return (
      <Fragment>
        {options &&
          <Fragment>
            <Divider />
            <ProductDetailOptionsList options={options} />
          </Fragment>
        }
      </Fragment>
    );
  }

  render() {
    const { classes: { variantsContainer }, variants } = this.props;
    return (
      <div className={variantsContainer}>
        {variants.map(this.renderVariant)}
        {this.renderOptionsList()}
      </div>
    );
  }
}

export default VariantList;
