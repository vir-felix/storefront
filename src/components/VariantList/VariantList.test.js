import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import VariantList from "./";

test("basic snapshot", () => {
  const component = renderer.create(shallow(<VariantList />).get(0));
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
