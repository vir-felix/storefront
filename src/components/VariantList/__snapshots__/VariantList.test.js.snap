// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot 1`] = `
<ul
  className="VariantList-variantsList-1"
>
  <li
    className="VariantList-variantItem-2"
  >
    <button
      className="MuiButtonBase-root-5 VariantItem-variantButton-3 VariantItem-activeVariant-4"
      disabled={undefined}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiTypography-root-8 MuiTypography-body1-17"
      >
        Sample Variant
      </span>
      <span
        className="MuiTypography-root-8 MuiTypography-body1-17"
      />
    </button>
  </li>
  <li
    className="VariantList-variantItem-2"
  >
    <button
      className="MuiButtonBase-root-5 VariantItem-variantButton-3"
      disabled={undefined}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiTypography-root-8 MuiTypography-body1-17"
      >
        Sample Variant 2
      </span>
      <span
        className="MuiTypography-root-8 MuiTypography-body1-17"
      />
    </button>
  </li>
</ul>
`;
